docs/examples/nmt/transformer.md

```diff
@@ -0,0 +1,698 @@
+                                                        bpe=False)
+
+print('WMT14 EN-DE SOTA model test loss: %.2f; test bleu score: %.2f; time cost %.2fs'
+      %(wmt_test_loss, wmt_test_bleu_score * 100, (time.time() - eval_start_time)))
+```
+
+```{.json .output n=16}
+[
+ {
+  "name": "stdout",
+  "output_type": "stream",
+  "text": "WMT14 EN-DE SOTA model test loss: 1.21; test bleu score: 27.35; time cost 782.25s\n"
+ }
+]
+```
+
+```{.python .input  n=17}
+print('Sample translations:')
+num_pairs = 3
+
+for i in range(num_pairs):
+    print('EN:')
+    print(wmt_test_text[i][0])
+    print('DE-Candidate:')
+    print(wmt_test_translation_out[i])
+    print('DE-Reference:')
+    print(wmt_test_tgt_sentences[i])
+    print('========')
+```
+
+```{.json .output n=17}
+[
+ {
+  "name": "stdout",
+  "output_type": "stream",
+  "text": "Sample translations:\nEN:\nOrlando Bloom and Miranda Kerr still love each other\nDE-Candidate:\nOrlando Bloom und Miranda Kerr lieben einander immer noch.\nDE-Reference:\nOrlando Bloom und Miranda Kerr lieben sich noch immer\n========\nEN:\nActors Orlando Bloom and Model Miranda Kerr want to go their separate ways.\nDE-Candidate:\nSchauspieler Orlando Bloom und Model Miranda Kerr wollen ihre eigenen Wege gehen.\nDE-Reference:\nSchauspieler Orlando Bloom und Model Miranda Kerr wollen k\u00fcnftig getrennte Wege gehen.\n========\nEN:\nHowever, in an interview, Bloom has said that he and Kerr still love each other.\nDE-Candidate:\nIn einem Interview hat Bloom jedoch gesagt, dass er und Kerr sich immer noch lieben.\nDE-Reference:\nIn einem Interview sagte Bloom jedoch, dass er und Kerr sich noch immer lieben.\n========\n"
+ }
+]
+```
+
+### Translation Inference
+
+We herein show the actual translation example (EN-DE) when given a source language using the SOTA Transformer model.
+
+```{.python .input  n=18}
+import utils
+
+print('Translate the following English sentence into German:')
+
+sample_src_seq = 'We love each other'
+
+print('[\'' + sample_src_seq + '\']')
+    
+sample_tgt_seq = utils.translate(wmt_translator, 
+                                 sample_src_seq, 
+                                 wmt_src_vocab, 
+                                 wmt_tgt_vocab, 
+                                 wmt_detokenizer,
+                                 ctx)
+
+print('The German translation is:')
+print(sample_tgt_seq)
+```
+
+```{.json .output n=18}
+[
+ {
+  "name": "stdout",
+  "output_type": "stream",
+  "text": "Translate the following English sentence into German:\n['We love each other']\nThe German translation is:\n['Wir lieben einander']\n"
+ }
+]
+```
+
+## Train Your Own Transformer
+
+In this subsection, we will go though the whole process about loading translation dataset in a more unified way, and create data sampler and loader, as well as define the Transformer model, finally writing training script to train the model yourself.
+
+### Load and Preprocess TOY Dataset
```

Comment:
How about not using the TOY Dataset at all, because it is not a 'well-known' dataset but rather only exists inside GluonNLP for the purpose of speeding up some tests on the CI server. How about loading the WMT2014BPE here but in case that `demo = True`, throwing away all but the first X% of the dataset away?

Reply:
Thanks for the suggestion @leezu . The reasons we are using TOY dataset instead of using WMT2014BPE are three folds: 1) The reason we are using TOY dataset is for showing how we can train the transformer in an unified data API. It is not trivial to use X% of the whole WMT2014BPE. We need to hack the data API, which will cause the API inconsistency compared to directly the data API usage in the   first portion of the notebook. 2) To hack the data API is not trivial: it requires getting the batch depends on the translation dataset transforms; similarly, creating the valid_lengths, samplers and data loader depends on the translation dataset. So we need to hack the translation dataset to use certain portion of the WMT dataset. Besides, we also need to hack the samplers and data loader. 3) Since as you know it is impossible to see the reasonable training result in the notebook. Despite the data loading issue, as you know it requires thousands of warm up steps. All in all, to show how to train the model, instead, simply using TOY dataset is enough and more reasonable.

Reply:
Thanks for the explanation @cgraywang . I'm not sure if I can completely follow though. As far as I understand, WMT2014BPE is simply a gluon.data.ArrayDataset containing source and translated strings. So to get a subset of WMT2014BPE the following works:
```
>>> d = nlp.data.WMT2014BPE()
>>> first_thousand_src, first_thousand_target = d[:1000]
>>> small_d = mx.gluon.data.ArrayDataset(first_thousand_src, first_thousand_target)
>>> # or short mx.gluon.data.ArrayDataset(*d[:1000])
```

As far as I understand, we only need to delete below case from the `load_translation_data` function:
```
    elif dataset == 'TOY':
        common_prefix = 'TOY_{}_{}_{}_{}'.format(src_lang, tgt_lang,
                                                 hparams.src_max_len, hparams.tgt_max_len)
        data_train = nmt.dataset.TOY('train', src_lang=src_lang, tgt_lang=tgt_lang)
        data_val = nmt.dataset.TOY('val', src_lang=src_lang, tgt_lang=tgt_lang)
        data_test = nmt.dataset.TOY('test', src_lang=src_lang, tgt_lang=tgt_lang) 
```

and replace it with `mx.gluon.data.ArrayDataset(*d[:1000])` for train val and test.

To the interface cleaner, it may be nice to introduce some parameter, e.g. `k` to load_translation_data that allows to specify how many sentences at maximum to load (i.e. 1000 above) instead of simply returning only 1000 sentences if `demo=True`.


Given that this directly changes `load_translation_data`, I don't see why there would be any issues with valid lengths, sampler etc, as all of that should now be automatically be constructed with respect to the shorter dataset (ie. your points 2) and 3))

I believe this change would make the notebook easier to understand as there is no need for the TOY dataset here and it may confuse people reading the code.

Reply:
Thanks @leezu for the quick response. I probably can make my explanation clearer... The purpose of the training script in the notebook is to show the users how to use a dataset (maybe user data, maybe not) to train their own transformer using the API already in the script. Then: 

1) if we are hacking the dataset the way you just showed, users will be confused, they probably will ask a) how to use my data set to train the model? will I need to split the dataset? do I need to change the code in the existing API (for example, `load_translation_data`)? This will break the existing API. 

2) The `TOY` dataset is essentially a sample from the WMT2014-EN-DE dataset. I show an example:

`data_train[0] = ('Orlando Bloom and Miranda Kerr still love each other', 'Orlando Bloom und Miranda Kerr lieben sich noch immer')`

3) If users turn on 'demo = False', it will take at least 5min to download the dataset and 1 day to make the training completed to get the SoTA results. If users intend to use the `demo` mode, the dataset should be small and show how to load the dataset without hacking the API. I guess we already agree on this point.

In conclusion, I believe use what we have right now is clear to the users.

Reply:
The download time is indeed a valid concern. For people with slower access to s3 (e.g., China), it might be a show stopper to force these users to sit there and wait for 30m just to get started.

On the other hand, our machine translation dataset already consists of different segments. For example:
https://github.com/dmlc/gluon-nlp/blob/master/gluonnlp/data/translation.py#L176-L192

It's conceivable that we can download one small segment for demonstration purpose. However, for the ease of implementation, we chose to package everything in a single file, which makes it impossible to download one single segment right now. I think the use-case in this PR is a reasonable justification to have per-segment download.

I will leave it to @cgraywang @leezu you two to decide how to proceed.

Reply:
I agree we should not download very large corpora for our example scripts. So for that reason I'm fine with using the toy corpus. I think it would be better to distribute large corpora in multiple files and have our API handle the case where only a subset of the dataset is requested. (But that is out of scope now). That would make the toy corpus obsolete.

Regarding point 1) and 2). I think on the contrary, coding up a new dataset class and doing the preprocessing manually (which is what is done for TOY), is not showing any strength of our dataset API. Instead it is nice to show how users can adapt the existing gluon-nlp library facilities to their needs.

Reply:
The item of breaking down the dataset is tracked in https://github.com/dmlc/gluon-nlp/issues/358

Reply:
Thanks @leezu @szha for the helpful discussion on the notebook. I agree that if the large dataset can be downloaded in segments, it would be better and we can remove the TOY dataset entirely.